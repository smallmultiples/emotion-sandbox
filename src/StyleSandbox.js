import React, { useMemo } from "react";
import PropTypes from "prop-types";
import { CacheProvider } from "@emotion/core";
import createSandboxPlugin from "./createSandboxPlugin";

const propTypes = {
  children: PropTypes.node.isRequired,
  elementId: PropTypes.string.isRequired,
};

function StyleSandbox({ children, elementId }) {
  const styleCache = useMemo(() => createSandboxPlugin(`#${elementId}`), [elementId]);

  return (
    <CacheProvider value={styleCache}>
      <div id={elementId}>
        {children}
      </div>
    </CacheProvider>
  );
}

StyleSandbox.propTypes = propTypes;

export default StyleSandbox;
