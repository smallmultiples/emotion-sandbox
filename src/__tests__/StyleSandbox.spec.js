import React from "react";
import styled from "@emotion/styled";
import renderer from "react-test-renderer";
import serializer from "jest-emotion";
import StyleSandbox from "../StyleSandbox";

expect.addSnapshotSerializer(serializer);

it("renders correctly", () => {
  const tree = renderer.create(
    <StyleSandbox elementId="myId">
      <div className="myClass">test</div>
    </StyleSandbox>,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it("renders styles correctly", () => {
  const Div = styled("div")({
    background: "white",
    color: "black",
  });

  const tree = renderer.create(
    <StyleSandbox elementId="myId">
      <Div>test</Div>
    </StyleSandbox>,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it("renders pseudoclasses correctly", () => {
  const A = styled("a")({
    color: "blue",
    "&:hover, &:focus": {
      color: "orange",
    },
    "&, &:hover, &:focus": {
      textDecoration: "underline",
    },
  });

  const tree = renderer.create(
    <StyleSandbox elementId="myId">
      <A>test</A>
    </StyleSandbox>,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it("renders pseudoelements correctly", () => {
  const Div = styled("div")({
    "&::before": {
      content: JSON.stringify("prefix"),
    },
    "&::after": {
      content: JSON.stringify("suffix"),
    },
  });

  const tree = renderer.create(
    <StyleSandbox elementId="myId">
      <Div>test</Div>
    </StyleSandbox>,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it("renders descendent elements correctly", () => {
  const Div = styled("div")({
    "& ~ span": {
      content: JSON.stringify("prefix"),
    },
    "a:hover + &": {
      content: JSON.stringify("suffix"),
    },
  });

  const tree = renderer.create(
    <StyleSandbox elementId="myId">
      <Div>test</Div>
    </StyleSandbox>,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
