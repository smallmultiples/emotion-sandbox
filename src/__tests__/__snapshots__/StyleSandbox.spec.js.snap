// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id="myId"
>
  <div
    className="myClass"
  >
    test
  </div>
</div>
`;

exports[`renders descendent elements correctly 1`] = `
#myId .emotion-0 ~ span {
  content: "prefix";
}

#myId a:hover + .emotion-0 {
  content: "suffix";
}

<div
  id="myId"
>
  <div
    className="emotion-0"
  >
    test
  </div>
</div>
`;

exports[`renders pseudoclasses correctly 1`] = `
#myId .emotion-0 {
  color: blue;
}

#myId .emotion-0:hover,
#myId .emotion-0:focus {
  color: orange;
}

#myId .emotion-0,
#myId .emotion-0:hover,
#myId .emotion-0:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  id="myId"
>
  <a
    className="emotion-0"
  >
    test
  </a>
</div>
`;

exports[`renders pseudoelements correctly 1`] = `
#myId .emotion-0::before {
  content: "prefix";
}

#myId .emotion-0::after {
  content: "suffix";
}

<div
  id="myId"
>
  <div
    className="emotion-0"
  >
    test
  </div>
</div>
`;

exports[`renders styles correctly 1`] = `
#myId .emotion-0 {
  background: white;
  color: black;
}

<div
  id="myId"
>
  <div
    className="emotion-0"
  >
    test
  </div>
</div>
`;
